Dockerfile

```diff
@@ -69,41 +73,50 @@ RUN freshclam --quiet
 
 # Create a3m user
 RUN set -ex \
-	&& groupadd --gid 333 --system a3m \
-	&& useradd --uid 333 --gid 333 --create-home --home-dir /home/a3m --system a3m \
+	&& groupadd --gid ${GROUP_ID} --system a3m \
+	&& useradd --uid ${USER_ID} --gid ${GROUP_ID} --create-home --home-dir /home/a3m --system a3m \
 	&& mkdir -p /home/a3m/.local/share/a3m \
 	&& chown -R a3m:a3m /home/a3m/.local
 
 
 #
-# Archivematica
+# a3m
 #
 
 FROM base AS a3m
 
+ARG PYTHON_VERSION=3.9
 ARG REQUIREMENTS=/a3m/requirements-dev.txt
 ARG DJANGO_SETTINGS_MODULE=a3m.settings.common
 ENV DJANGO_SETTINGS_MODULE=${DJANGO_SETTINGS_MODULE}
+ENV PATH=/home/a3m/.local/a3m-venv/bin:$PATH
 
 COPY ./a3m/externals/fido/ /usr/lib/archivematica/archivematicaCommon/externals/fido/
 COPY ./a3m/externals/fiwalk_plugins/ /usr/lib/archivematica/archivematicaCommon/externals/fiwalk_plugins/
 
 RUN set -ex \
 	&& add-apt-repository ppa:deadsnakes/ppa \
 	&& apt-get update \
-	&& apt-get install -y --no-install-recommends python3.9 python3.9-distutils build-essential libpython3.9-dev \
-	&& update-alternatives --install /usr/bin/python python /usr/bin/python3.9 1 \
-	&& update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 1 \
-	&& curl -q https://bootstrap.pypa.io/get-pip.py | python \
+	&& apt-get install -y --no-install-recommends \
+		build-essential \
+		python${PYTHON_VERSION} \
+		python${PYTHON_VERSION}-venv \
+		python${PYTHON_VERSION}-distutils \
+		libpython${PYTHON_VERSION}-dev \
+	&& update-alternatives --install /usr/bin/python python /usr/bin/python${PYTHON_VERSION} 1 \
+	&& update-alternatives --install /usr/bin/python3 python3 /usr/bin/python${PYTHON_VERSION} 1 \
 	&& rm -rf /var/lib/apt/lists/*
 
+USER a3m
+
 COPY ./requirements.txt /a3m/requirements.txt
 COPY ./requirements-dev.txt /a3m/requirements-dev.txt
-RUN python -m pip install -r ${REQUIREMENTS}
+RUN set -ex \
+	&& python -m venv /home/a3m/.local/a3m-venv \
+	&& pip install --upgrade pip \
+	&& pip install --no-cache-dir -r ${REQUIREMENTS}
```

Comment:
Why a virtual env. inside the container?

Reply:
I know the container already provides a layer of isolation, but this also avoids clashes with Ubuntu packages that may depend on the default/system Python in case [custom images](https://a3m.readthedocs.io/en/latest/docker.html#custom-images) are based on this.

And even though it would probably not be possible yet for `a3m` because of all the MCPClient related dependencies (`ffmpeg`, `ghostscript`, etc), this [Multi-stage builds #2: Python specifics](https://pythonspeed.com/articles/multi-stage-docker-python/#solution-2-virtualenv) article explains how you could go a bit further and pass only the venv directory to a following build stage.

Reply:
Thanks for the explanation! :ok_hand: